[WPE] Drop ALLOW_UNSAFE_BUFFER_USAGE from PlatformPasteboardLibWPE

csaavedra opened the pull request:
#### c7fe0332292c1e865c6133f9c18e38317e18bdca
<pre>
[WPE] Drop ALLOW_UNSAFE_BUFFER_USAGE from PlatformPasteboardLibWPE
<a href="https://bugs.webkit.org/show_bug.cgi?id=283748">https://bugs.webkit.org/show_bug.cgi?id=283748</a>

Reviewed by Adrian Perez de Castro.

Prefer ASCIILiteral and std:array to raw C types.

* Source/WebCore/platform/libwpe/PlatformPasteboardLibWPE.cpp:
(WebCore::PlatformPasteboard::getTypes const):
(WebCore::PlatformPasteboard::write):

Canonical link: <a href="https://commits.webkit.org/287123@main">https://commits.webkit.org/287123@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c1500d7bfc9e27add351740f80090f738c7a9b1d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78438 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57483 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/31820 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83099 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29703 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/80571 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66634 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5764 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61423 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19340 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81505 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51435 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/69191 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41736 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48782 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/25177 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28040 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69878 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/25548 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84465 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5803 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/3951 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69648 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5964 "Built successfully") | [⏳ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/macOS-Release-WK2-Stress-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68903 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12926 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12119 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5750 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/8766 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5739 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9172 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7525 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebCore/platform/libwpe/PlatformPasteboardLibWPE.cpp:
```diff
@@ -84,23 +80,21 @@ String PlatformPasteboard::readString(size_t, const String& type) const
 
-    WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN // WPE port
-    struct wpe_pasteboard_string_pair pairs[] = {
+    std::array<struct wpe_pasteboard_string_pair, 2> pairs = { {
         { { nullptr, 0 }, { nullptr, 0 } },
         { { nullptr, 0 }, { nullptr, 0 } },
```
Using  `wpe_pasteboard_string_initialize()` copies the strings, then the implementation behind `wpe_pasteboard_write()` will make a second copy, and then each string is freed below with `wpe_pasteboard_string_free()`.

We could initialize the items here with the initializer list, and skip the `wpe_pasteboard_string_free()` call below. Also, notice that there is an `ASCIILiteral::length()` function, we do not need to call `strlen()` manually:

```suggestion
    std::array<struct wpe_pasteboard_string_pair, 2> pairs = { {
        { { plainText, plainText.length() }, { textString.data(), textString.length() } },
        { { htmlText, htmlText.length() }, { markupString.data(), markupString.length() } },
```

csaavedra replied:
This doesn't work because `ASCIILiteral` will implicitly convert to a `const char*` and it seems that a `const_cast` to a `char*` is disallowed. For the `CString` part one needs to use `::mutableData()`, but since one can't use the ASCIILiteral directly, I suspect this won't be possible.

aperezdc replied:
Fair, I didn't remember that the byte array in `wpe_pasteboard_string` is `char*`, without a `const`. Let's leave it as-is, then.

aperezdc commented on Source/WebCore/platform/libwpe/PlatformPasteboardLibWPE.cpp:
```diff
@@ -84,23 +80,21 @@ String PlatformPasteboard::readString(size_t, const String& type) const
     wpe_pasteboard_string_initialize(&pairs[1].type, htmlText, strlen(htmlText));
     wpe_pasteboard_string_initialize(&pairs[1].string, markupString.data(), markupString.length());
-    struct wpe_pasteboard_string_map map = { pairs, 2 };
+    struct wpe_pasteboard_string_map map = { pairs.data(), 2 };
```
I would use `pairs.size()` to have the `2` only in the declaration of the `std::array` and making the intention clearer:

```suggestion
    struct wpe_pasteboard_string_map map = { pairs.data(), pairs.size() };
```

aperezdc commented on Source/WebCore/platform/libwpe/PlatformPasteboardLibWPE.cpp:
```diff
@@ -84,23 +80,21 @@ String PlatformPasteboard::readString(size_t, const String& type) const
-    static const char plainText[] = "text/plain;charset=utf-8";
-    static const char htmlText[] = "text/html;charset=utf-8";
+    static const auto plainText = "text/plain;charset=utf-8"_s;
+    static const auto htmlText = "text/html;charset=utf-8"_s;
```
These could even be `constexpr`:

```suggestion
    static constexpr auto plainText = "text/plain;charset=utf-8"_s;
    static constexpr auto htmlText = "text/html;charset=utf-8"_s;
```

aperezdc reviewed (changes requested):
I left a couple of comments, PTAL.

csaavedra committed:
[WPE] Drop ALLOW_UNSAFE_BUFFER_USAGE from PlatformPasteboardLibWPE
https://bugs.webkit.org/show_bug.cgi?id=283748

Reviewed by Adrian Perez de Castro.

Prefer ASCIILiteral and std:array to raw C types.

* Source/WebCore/platform/libwpe/PlatformPasteboardLibWPE.cpp:
(WebCore::PlatformPasteboard::getTypes const):
(WebCore::PlatformPasteboard::write):

Canonical link: https://commits.webkit.org/287123@main
